### PR TITLE
Resnest

### LINKED ISSUE #505: Add ResNeSt
https://hangzhang.org/files/resnest.pdf

### PR DESCRIPTION
**Pull request by anton-br**
this pr is adding a model form https://arxiv.org/abs/2004.08955

Fixes #505

### CODE REVIEW

**Review by SergeyTsimfer — COMMENTED** [review]
It is virtually impossoble to read changed files now due to `CodeCov` outputs
@roman-kh is that intended behaviour?
@anton-br please, add notebook (like [this one](https://github.com/analysiscenter/batchflow/blob/master/examples/experiments/self_attention/research_self_attention.ipynb)) to  showcase the model performance with this block

File: batchflow/models/torch/blocks.py [hunk 1]
```diff
@@ -298,3 +298,66 @@ def __init__(self, inputs=None, layout='nacd', filters=None, kernel_size=3, stri
...
+            branch_params = {'layout': 'cn', 'filters': filters,
+                             'kernel_size': 1, 'strides': strides}
+        else:
+            branch_params = {}
+        layout = 'R' + layout + op
+
+        layer_params = [{'branch': branch_params,
+                         'branch/strides': strides}]
+
+        layer_params += [{}]*(n_reps-1)
```
Comment (SergeyTsimfer) [inline batchflow/models/torch/blocks.py:356]: What about the `n_repeats` parameter of `ConvBlock`?

**Review by SergeyTsimfer — COMMENTED** [review]
Add brief description of the main idea for each block

**Comment by roman-kh** [comment]
> It is virtually impossible to read changed files now due to `CodeCov` outputs
> @roman-kh is that intended behaviour?
I believe this can be changed via codecov configuration file. I've no idea how to do this.

File: batchflow/models/torch/blocks.py [hunk 2]
```diff
@@ -298,3 +298,84 @@ def __init__(self, inputs=None, layout='nacd', filters=None, kernel_size=3, stri
     def forward(self, x):
         x = super().forward(x)
         return x if self.skip else x[:, self.input_num_channels:]
+
+
+class ResNeStBlock(ConvBlock):
+    """ ResNeSt Module: pass tensor through one or multiple (`n_reps`) blocks, each of which is a
```
Comment (SergeyTsimfer) [inline batchflow/models/torch/blocks.py:304]: Make a clear description of what happens to the `input` tensor:

- first of all, we split it into `radix` groups
- then ...
- then apply `sac`
- then ...

File: batchflow/models/torch/blocks.py [hunk 3]
```diff
@@ -298,3 +298,84 @@ def __init__(self, inputs=None, layout='nacd', filters=None, kernel_size=3, stri
...
+        Convolution kernel size. Default is 3.
+    radix : int
+        The number of splits within a cardinal group. Default is 2.
+    cardinality : int
+        The number of feature-map groups. Given feature-map is splitted to groups with same size. Default is 1.
+    strides : int, list of int
+        Convolution stride. Default is 1.
+    reduction_factor : int
+        The number reflecting the size of the filter reduction in the inner layer. Default is 1.
+    bottleneck_width : int
```
Comment (SergeyTsimfer) [inline batchflow/models/torch/blocks.py:339]: `reduction_factor` and `bottleneck_width` are actually doing the same: one inside `SAC`, and one inside this block directly.

File: batchflow/models/torch/blocks.py [hunk 4]
```diff
@@ -298,3 +298,84 @@ def __init__(self, inputs=None, layout='nacd', filters=None, kernel_size=3, stri
...
+        The size of the reduction in the number of filters in the entire block. Default is 64.
+    attention : str
```
Comment (SergeyTsimfer) [inline batchflow/models/torch/blocks.py:341]: Should always be `sac`, as it is an essential part of the module

File: batchflow/models/torch/blocks.py [hunk 5]
```diff
@@ -298,3 +298,84 @@ def __init__(self, inputs=None, layout='nacd', filters=None, kernel_size=3, stri
...
+                         'branch/strides': strides}]
+
+        layer_params += [{}]*(n_reps-1)
+
+        # All given parameters are going directly to attention module due to the lack of other operations in the block.
+        self_attention = dict(radix=radix, cardinality=cardinality, kernel_size=kernel_size,
+                              filters=block_filters, strides=strides, reduction_factor=reduction_factor,
+                              external_mult=external_mult)
+
+        super().__init__(*layer_params, inputs=inputs, layout=layout, attention=attention,
```
Comment (SergeyTsimfer) [inline batchflow/models/torch/blocks.py:379]: Remove `strides`

File: batchflow/models/torch/blocks.py [hunk 6]
```diff
@@ -298,3 +298,84 @@ def __init__(self, inputs=None, layout='nacd', filters=None, kernel_size=3, stri
...
+            branch_params = {}
+        layout = 'R' + layout + op
+
+        layer_params = [{'branch': branch_params,
+                         'branch/strides': strides}]
+
+        layer_params += [{}]*(n_reps-1)
+
+        # All given parameters are going directly to attention module due to the lack of other operations in the block.
+        self_attention = dict(radix=radix, cardinality=cardinality, kernel_size=kernel_size,
```
Comment (SergeyTsimfer) [inline batchflow/models/torch/blocks.py:375]:
```python
attention_kwargs = {
...
}
```

File: batchflow/models/torch/blocks.py [hunk 7]
```diff
@@ -298,3 +298,84 @@ def __init__(self, inputs=None, layout='nacd', filters=None, kernel_size=3, stri
...
+                              filters=block_filters, strides=strides, reduction_factor=reduction_factor,
+                              external_mult=external_mult)
```
Comment (SergeyTsimfer) [inline batchflow/models/torch/blocks.py:377]: `external_mult` is actually not used in `sac`. One more reason to remove it

File: batchflow/models/torch/blocks.py [hunk 8]
```diff
@@ -298,3 +298,84 @@ def __init__(self, inputs=None, layout='nacd', filters=None, kernel_size=3, stri
...
+    op : str or callable
+        Operation for combination shortcut and residual.
+        See more :class:`~.layers.Combine` documentation. Default is '+a'.
+    n_reps : int
+        Number of times to repeat the whole block. Default is 1.
+    kwargs : dict
+        Other named arguments for the :class:`~.layers.ConvBlock`.
+    """
+    def __init__(self, inputs=None, layout='cnaScn', filters='same', kernel_size=3, radix=2, cardinality=1,
+                 strides=1, reduction_factor=1, external_mult=4, bottleneck_width=64, attention='sac', op='+a',
```
Comment (SergeyTsimfer) [inline batchflow/models/torch/blocks.py:353]: Undocumented `external_mult`, which suddenly changes the amount of desired filters: say, I passed `filters=64`, but for some reason, I get the tensor with `256` channels...

File: batchflow/models/torch/layers/activations.py [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+    Returns
+    -------
+    x : torch Tensor
+        The output size will be (batch size, `radix`).
+
+    Note
+    ----
+    If `radix` is 1, common sigmoid is used.
+    """
+    def __init__(self, radix, cardinality, add_dims=0):
```
Comment (SergeyTsimfer) [inline batchflow/models/torch/layers/activations.py:30]: Undocumented arguement

File: batchflow/models/torch/layers/activations.py [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+
+    def forward(self, x):
+        batch = x.size(0)
+        if self.radix > 1:
+            x = x.view(batch, self.cardinality, self.radix, -1).transpose(1, 2)
+            x = torch.nn.functional.softmax(x, dim=1)
+            x = x.reshape(batch, -1)
+        else:
+            x = torch.sigmoid(x)
+        if self.add_dims > 0:
```
Comment (SergeyTsimfer) [inline batchflow/models/torch/layers/activations.py:44]: Would much rather use `>` letter in the `ConvBlock`

File: batchflow/models/torch/layers/attention.py [hunk 11]
```diff
@@ -547,3 +558,101 @@ def __repr__(self):
...
+        else:
+            concatted = x
+        att = self.avgpool_conv1d(concatted)
+        att = self.rsoftmax(att)
+        if self.radix > 1:
+            attens = torch.split(att, rchannel//self.radix, dim=1)
+            result = sum([att*split for (att, split) in zip(attens, splitted)])
+        else:
+            result = att * x
+        return result.contiguous()
```
Comment (SergeyTsimfer) [inline batchflow/models/torch/layers/attention.py:650]: Test if you really need it

**Review by SergeyTsimfer — CHANGES_REQUESTED** [review]

File: batchflow/models/torch/layers/__init__.py [hunk 12]
```diff
@@ -1,11 +1,12 @@
...
 from .conv_block import BaseConvBlock, ConvBlock, update_layers
-from .core import Flatten, Dense, Activation, Dropout, AlphaDropout, BatchNorm
+from .core import Flatten, Dense, Dropout, AlphaDropout, BatchNorm
 from .conv import Conv, ConvTranspose, DepthwiseConv, DepthwiseConvTranspose, SeparableConv, SeparableConvTranspose
 from .pooling import MaxPool, AvgPool, Pool, AdaptiveMaxPool, AdaptiveAvgPool, AdaptivePool, \
 					 GlobalPool, GlobalMaxPool, GlobalAvgPool, ChannelPool
 from .resize import IncreaseDim, Reshape, Interpolate, PixelShuffle, SubPixelConv, \
 					Upsample, Combine, Crop
 from .attention import SelfAttention, SEBlock, SCSEBlock, SimpleSelfAttention, BAM, CBAM, FPA, SelectiveKernelConv
+from .activation import Activation
```
Comment (SergeyTsimfer) [inline batchflow/models/torch/layers/__init__.py:11]: `activation` should be before some other stuff, right after `core`

File: batchflow/models/torch/layers/attention.py [hunk 13]
```diff
@@ -547,3 +557,110 @@ def __repr__(self):
...
+    * Then attention takes place:
+        * Here, we summarize feature maps by groups and apply Global Average Pooling.
+        * Then we apply two 1x1 convolutions with groups=`cardinality`. The number of filters in the first
+            convolution is `filters`*`radix` // `reduction_factor`.
+        * Then we use RadixSoftmax, which applies a softmax for feature maps grouped into `cardinality` groups.
+        * Then, the resulting groups are summed up with feature maps before the attention part.
+
+    Parameters
+    ----------
+    kernel_size : int or list of int
```
Comment (SergeyTsimfer) [inline batchflow/models/torch/layers/attention.py:581]: You don't actually need this parameter; `kwargs` can take care of it

File: batchflow/models/torch/layers/conv_block.py [hunk 14]
```diff
@@ -403,8 +404,13 @@ def __getitem__(self, key):
     def __repr__(self):
         if getattr(self, 'short_repr', False):
             msg = f'layout={self.layout}\n'
-            msg += '\n'.join([f'{key}' for key in self.keys()])
-            return f'{msg}\n'
+            # msg += '\n'.join([f'{key}' for key in self.keys()])
+
+            for i, (key, value) in enumerate(self.items()):
```
Comment (SergeyTsimfer) [inline batchflow/models/torch/layers/conv_block.py:409]: Don't include that into PR

File: batchflow/models/torch/blocks.py [hunk 15]
```diff
@@ -298,3 +298,98 @@ def __init__(self, inputs=None, layout='nacd', filters=None, kernel_size=3, stri
     def forward(self, x):
         x = super().forward(x)
         return x if self.skip else x[:, self.input_num_channels:]
+
+
+class ResNeStBlock(ConvBlock):
+    """ ResNeSt Module: pass tensor through one or multiple (`n_reps`) blocks, each of which is processed the input
```
Comment (SergeyTsimfer) [inline batchflow/models/torch/blocks.py:304]: `ResNeSt Module: apply following operations to a supplied tensor:`

File: batchflow/models/torch/blocks.py [hunk 16]
```diff
@@ -298,3 +298,98 @@ def __init__(self, inputs=None, layout='nacd', filters=None, kernel_size=3, stri
...
+    tensor as follows:
+    * First of all, it goes through 1x1 convolution with normalization and activation.
```
Comment (SergeyTsimfer) [inline batchflow/models/torch/blocks.py:306]: Missing indentation level

File: batchflow/models/torch/blocks.py [hunk 17]
```diff
@@ -298,3 +298,98 @@ def __init__(self, inputs=None, layout='nacd', filters=None, kernel_size=3, stri
...
+    * Then `sac` is applied:
```
Comment (SergeyTsimfer) [inline batchflow/models/torch/blocks.py:307]: `Split Attention Convolution`

File: batchflow/models/torch/blocks.py [hunk 18]
```diff
@@ -298,3 +298,98 @@ def __init__(self, inputs=None, layout='nacd', filters=None, kernel_size=3, stri
...
+            * Then we use RadixSoftmax, which applies a softmax for feature maps grouped into `radix` groups.
+            * Then, the resulting groups are summed up with feature maps before the attention part.
+    * The last layer of the block is a 1x1 convolution that increases the feature map from `filters` to
+      `filters`*`scaling_factor`.
+    * In the end, skip connection applies to the result of the ResNeStBlock.
+
+    The number of filters inside ResNeSt Attention calculates as following:
+    >>> filters = int(filters // reduction_factor) * cardinality
+
+    The implementation was inspired by the authors' code (`<https://github.com/zhanghang1989/ResNeSt>`_), thus
```
Comment (SergeyTsimfer) [inline batchflow/models/torch/blocks.py:324]: `is inspired`

**Review by SergeyTsimfer — APPROVED** [review]

**Review by SergeyTsimfer — APPROVED** [review]

**Pull request merged**